url39209c0557_au_research_ANDS_RIFCS_API_tgz-pJ8-org_ands_rifcs_example_SampleRIFCSBuildCollectionV1Dot6Number1J8/src/org/ands/rifcs/base/RelatedObject.java

```diff
@@ -0,0 +1,171 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+
+/**
+ * Class representing related object information.
+ *
+ * @author Scott Yeadon
+ *
+ */
+public class RelatedObject extends RIFCSElement {
+    /** List of Relation nodes. */
+    private List<Relation> relations = new ArrayList<Relation>();
+
+    /**
+     * Construct a RelatedObject object.
+     *
+     * @param n
+     *        A w3c Node, typically an Element
+     *
+     * @throws RIFCSException A RIFCSException
+     */
+    protected RelatedObject(final Node n) throws RIFCSException {
+        super(n, Constants.ELEMENT_RELATED_OBJECT);
+        initStructures();
+    }
+
+
+    /**
+     * Set the related object key value.
+     *
+     * @param keyValue
+     *      The key uniquely identifying the related registry object
+     */
+    public final void setKey(final String keyValue) {
+        Element key = this.newElement(Constants.ELEMENT_KEY);
+        key.setTextContent(keyValue);
+        this.getElement().appendChild(key);
+    }
+
+
+    /**
+     * Get the related object key value.
+     *
+     * @return
+     *     The key uniquely identifying the related registry object
+     */
+    public final String getKey() {
+        NodeList nl = super.getElements(Constants.ELEMENT_KEY);
+        if (nl.getLength() == 1) {
+            return nl.item(0).getTextContent();
+        }
+
+        return null;
+    }
+
+
+    /**
+     * Create and return an empty Relation object.
+     *
+     * The returned object has no properties or content and is not part
+     * of the RIF-CS document, it is essentially a constructor of an object
+     * owned by the RIF-CS document. The returned object needs to be
+     * "filled out" (e.g. with properties, additional sub-elements, etc)
+     * before being added to the RIF-CS document.
+     *
+     * @return the new Relation object
+     *
+     * @throws RIFCSException A RIFCSException
+     *
+     */
+    public final Relation newRelation() throws RIFCSException {
+        return new Relation(this.newElement(Constants.ELEMENT_RELATION));
+    }
+
+
+    /**
+     * Add a related object relation.
+     *
+     * @param type
+     *          The type of relation being described
+     * @param url
+     *    A URL expressing or implementing the relationship
+     *    between registry objects
+     * @param description
+     *    String describing the relation or null
+     * @param descriptionLanguage
+     *    The xml:lang attribute value
+     * @throws RIFCSException A RIFCSException
+     */
+    public final void addRelation(final String type,
+                            final String url,
+                            final String description,
+                            final String descriptionLanguage)
+                                    throws RIFCSException {
+        Relation relation = newRelation();
+
+        relation.setType(type);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;type&#39; (index: 0 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;description&#39;, &#39;url&#39;, &#39;descriptionLanguage&#39;. The same parameter type is validated here for example: url39209c0557_au_research_ANDS_RIFCS_API_tgz-pJ8-org_ands_rifcs_example_SampleRIFCSBuildCollectionV1Dot6Number1J8/src/org/ands/rifcs/base/DateElement.java:53. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).